---
title: "differential_motif"
author: "Nate Mullin"
date: "12/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Signac)
library(Seurat)
library(patchwork)
library(tidyverse)
library(EnsDb.Hsapiens.v86)
library(Matrix)
library(SummarizedExperiment)
library(ggpubr)
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(patchwork)
library(EnhancedVolcano)
set.seed(1234)
```


load data 
```{r}
 load("~/LSS/IVR/nkmullin/experiments/sc-atac-seq/melas_jci_insight_revision/motif/retina_trans_labels_annotated_na_removed_chromvar_12072022.RData")

load("~/LSS/IVR/nkmullin/experiments/sc-atac-seq/melas_jci_insight_revision/motif/rpechor_trans_labels_annotated_na_removed_chromvar_12072022.RData")

```


```{r}
# Plotting Loop
# retina
DefaultAssay(retina.atac.no.na) <- 'chromvar'
Idents(retina.atac.no.na) <- "celltype"

require(lattice)
for (i in levels(x = retina.atac.no.na)) {
  # get marker table
    tryCatch({
  output <- FindMarkers(retina.atac.no.na,
                                 only.pos=FALSE,
                                 min.pct= 0,
                                 logfc.threshold=0,
                                 test.use='LR',
                                 latent.vars='nCount_ATAC',
                                 ident.1 = "MELAS",ident.2 = "control",
                                 subset.ident = i, group.by = "disease")
  
  output$dpct <- output$pct.1 - output$pct.2
  output$gene <- rownames(output)
  
  for (j in 1:length(output$gene)) {
  output$motif[j] <- getMatrixByID(JASPAR2020, ID = output$gene[j])@name
     } 
  

  write.csv(x = output, file = paste0(i,"_diff_motif.csv"))
  
  fc_dpct <- ggplot(output, aes(x = dpct, y = avg_log2FC, label = motif)) +
                                  geom_point() +
                                  geom_label_repel(data=subset(output, abs(dpct) > 0.5 & abs(avg_log2FC) > 2), max.overlaps = 25)
      
  volcano  <-   EnhancedVolcano(output,
                                               lab = output$motif,
                                               x = 'avg_log2FC',
                                               y = 'p_val_adj',  
                                               title = i,
                                               cutoffLineWidth = 0.8,
                                               colAlpha = 1,
                                               pCutoff = 10e-6,
                                               FCcutoff = 2,
                                               pointSize = 2.0,
                                               labSize = 5.0,
                                               col=c('black', 'black', 'black', 'red3'),
                                               cutoffLineCol = 'dark grey',    
                                               gridlines.major = FALSE,
                                               gridlines.minor = FALSE) + NoLegend()
                                                 

  pdf(file = paste0(i,"_motif_fc_pct.pdf"), width = 5, height = 5)
    print(fc_dpct)
  dev.off()
  
  
  pdf(file = paste0(i,"_motif_volcano.pdf"), width = 8, height = 8)
    print(volcano)
  dev.off()
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}  # ERROR CELL TYPE ONLY WITH CELLS IN BOTH DISEASE STATES!

# rpe/choroid
DefaultAssay(rpechor.atac.no.na) <- 'chromvar'
Idents(rpechor.atac.no.na) <- "celltype"

for (i in levels(x = rpechor.atac.no.na)) {
  # get marker table
   tryCatch({
  output <- FindMarkers(rpechor.atac.no.na,
                                 only.pos=FALSE,
                                 min.pct= 0,
                                 logfc.threshold=0,
                                 test.use='LR',
                                 latent.vars='nCount_ATAC',
                                 ident.1 = "MELAS",ident.2 = "control",
                                 subset.ident = i, group.by = "disease")
  
  output$dpct <- output$pct.1 - output$pct.2
  output$gene <- rownames(output)
  
  for (j in 1:length(output$gene)) {
  output$motif[j] <- getMatrixByID(JASPAR2020, ID = output$gene[j])@name
     } 
  

  write.csv(x = output, file = paste0(i,"_diff_motif.csv"))
  
  fc_dpct <- ggplot(output, aes(x = dpct, y = avg_log2FC, label = motif)) +
                                  geom_point() +
                                  geom_label_repel(data=subset(output, abs(dpct) > 0.5 & abs(avg_log2FC) > 2), max.overlaps = 25)
      
  volcano  <-   EnhancedVolcano(output,
                                               lab = output$motif,
                                               x = 'avg_log2FC',
                                               y = 'p_val_adj',  
                                               title = i,
                                               cutoffLineWidth = 0.8,
                                               colAlpha = 1,
                                               pCutoff = 10e-6,
                                               FCcutoff = 1,
                                               pointSize = 2.0,
                                               labSize = 5.0,
                                               col=c('black', 'black', 'black', 'red3'),
                                               cutoffLineCol = 'dark grey',    
                                               gridlines.major = FALSE,
                                               gridlines.minor = FALSE) + NoLegend()
                                                 

  pdf(file = paste0(i,"_motif_fc_pct.pdf"), width = 5, height = 5)
    print(fc_dpct)
  dev.off()
  
  
  pdf(file = paste0(i,"_motif_volcano.pdf"), width = 8, height = 8)
    print(volcano)
  dev.off()
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
} 

```


# footprinting...

```{r, fig.width=5, fig.height=5}
# retina.atac.no.na <- Footprint(
#   object = retina.atac.no.na,
#   motif.name = c("ELK1", "ELK3", "ETV3", "KLF9", "KLF10", "FOXK2", "ZBTB12", "KLF9", "FOXP3", "FEV", "PHOX2A", "YY2", "GSX1", "LHX2"),
#   genome = BSgenome.Hsapiens.UCSC.hg38, assay = "ATAC")
# 
# p2 <- PlotFootprint(retina.atac.no.na, features = c("ELK1", "ELK3", "ETV3", "KLF9", "KLF10", "FOXK2", "ZBTB12", "KLF9", "FOXP3", "FEV", "PHOX2A", "YY2", "GSX1", "LHX2"), assay = "ATAC", group.by = "disease")
#  
#  pdf(file = "retina_motif_footprints_disease.pdf", width = 15, height = 15)
# p2
#   dev.off()


rpechor.atac.no.na <- Footprint(
  object = rpechor.atac.no.na,
  motif.name = c("YY1", "RFX2", "NR3C2"),
  genome = BSgenome.Hsapiens.UCSC.hg38, assay = "ATAC")




p1 <- PlotFootprint(rpechor.atac.no.na, features = c("YY1", "RFX2", "NR3C2"),
                    assay = "ATAC", group.by = "celltype", idents = c("Endothelial", "T_Cell", "Fibroblast"))

p1 + patchwork::plot_layout(ncol = 1)

p2 <- FeaturePlot(
  object = rpechor.atac.no.na,
  features = "MA0095.2",
  min.cutoff = 'q10',
  max.cutoff = 'q90',
  pt.size = 0.1, split.by = "disease"
)

table(rpechor.atac.no.na$celltype)

pdf("YY1_motif_fp.pdf", width = 5, height = 5)
PlotFootprint(rpechor.atac.no.na, features = "YY1",
              assay = "ATAC",
              group.by = "celltype",
              idents = c("T_Cell", "Macrophage", "B_Cell", "Schwann"))
dev.off()


pdf("NR3C2_motif_fp.pdf", width = 5, height = 5)
PlotFootprint(rpechor.atac.no.na, features = c("NR3C2"),
              assay = "ATAC", 
              group.by = "celltype",
              idents = c("Schwann", "Pericyte_SMC", "Endothelial"))
dev.off()

pdf("RFX2_motif_fp.pdf", width = 5, height = 5)
PlotFootprint(rpechor.atac.no.na, features = c("RFX2"),
              assay = "ATAC", 
              group.by = "celltype",
              idents = c("T_cell", "Endothelial"))
dev.off()

pdf("YY1_motif_vln.pdf", width = 5, height = 5)
VlnPlot(
  object = rpechor.atac.no.na,
  features = "MA0095.2",
  pt.size = 0.1, split.by = "disease"
)
dev.off()


pdf("NR3C2_motif_vln.pdf", width = 5, height = 5)
VlnPlot(
  object = rpechor.atac.no.na,
  features = "MA0727.1",
  pt.size = 0.1, split.by = "disease"
)
dev.off()

pdf("RFX2_motif_vln.pdf", width = 5, height = 5)
VlnPlot(
  object = rpechor.atac.no.na,
  features = "MA0600.2",
  pt.size = 0.1, split.by = "disease"
)
dev.off()



DimPlot(rpechor.atac.no.na) / p2 / p3



```
save rpechor object w footprint data
```{r}
save(rpechor.atac.no.na, file="rpechor_trans_labels_annotated_na_removed_chromvar_12072022.RData")

```

