---
title: "melas_wnn_analysis"
author: "Nate Mullin"
date: "1/11/2023"
output: html_document
---

```{r}
library(Seurat)
library(Signac)
library(tidyverse)
library(ggpubr)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggrepel)
library(RColorBrewer)
library(scCustomize)
```


###########################################
#### Run WNN to integrate RNA and ATAC ####
###########################################
```{r, fig.width=18, fig.height=6}
# load CCA output
load("~/LSS/IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cca__retina_01112023/final_object.RData")
# load LSI/Signac output
load("~/LSS/IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/lsi_retina_01112022/integrated_final.RData")

CCA_object <- IVR_object.combined
LSI_object <- integrated

# add correct fragment paths to atac object
FRAG_PATH1 <- "~/LSS/IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/20_009_rpechor/outs/atac_fragments.tsv.gz"
FRAG_PATH2 <- "~/LSS/IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/21_016_rpechor/outs/atac_fragments.tsv.gz"
FRAG_PATH3 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/21_016_rpeonly/outs/atac_fragments.tsv.gz"
FRAG_PATH4 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/21_028_rpeonly/outs/atac_fragments.tsv.gz"
FRAG_PATH5 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S1_21_016_per_retina/outs/atac_fragments.tsv.gz"
FRAG_PATH6 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S2_21_016_per_rpechor/outs/atac_fragments.tsv.gz"
FRAG_PATH7 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S3_20_058_per_retina/outs/atac_fragments.tsv.gz"
FRAG_PATH8 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S4_20_058_per_rpechor/outs/atac_fragments.tsv.gz"
FRAG_PATH9 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S5_20_067_per_retina/outs/atac_fragments.tsv.gz"
FRAG_PATH10 <- "~/LSS//IVR/nkmullin/experiments/multiome/melas_jci_insight_revision/cellranger_arc_runs/S6_20_067_per_rpechor/outs/atac_fragments.tsv.gz"



FRAG_PATHS <- c(FRAG_PATH5, FRAG_PATH7, FRAG_PATH9)
NEW_FRAG_LIST <- vector(length = 3, mode = "list")
for(i in 1:3){
  
  NEW_FRAG_LIST[[i]] <- UpdatePath(object = Fragments(LSI_object[["ATAC"]])[[i]],
                                   new.path = FRAG_PATHS[i],
                                   verbose = TRUE)
}

Fragments(LSI_object[["ATAC"]]) <- NULL
Fragments(LSI_object[["ATAC"]]) <- NEW_FRAG_LIST

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations)
# change to UCSC style since the data was mapped to hg19
genome(annotations) <- "GRCh38"
seqlevelsStyle(annotations) <- 'UCSC'

# add the gene information to the object
Annotation(LSI_object) <- annotations

# subset objects to only contain common cells that pass ATAC and RNA QC
CCA_object_subset <- subset(CCA_object, cells = colnames(LSI_object))
LSI_object_subset <- subset(LSI_object, cells = colnames(CCA_object_subset))

# set default assay
DefaultAssay(CCA_object_subset) <- "RNA"
DefaultAssay(LSI_object_subset) <- "ATAC"


# compare number of cells in each object
CCA_object # 26k cells
LSI_object # 22k cells
CCA_object_subset # 17k  cells
LSI_object_subset # 17k cells

# good numbers!

```


Create new multimodal object out of subsets
```{r}
# create a new object with relevant assays
multimodal_object <- CreateSeuratObject(counts =  CCA_object_subset@assays$RNA@counts,assay = "RNA")
multimodal_object[["integrated"]] <- CCA_object_subset@assays$integrated
multimodal_object[["ATAC"]] <- LSI_object_subset[["ATAC"]]

# and metadata
multimodal_object@meta.data <- LSI_object_subset@meta.data
multimodal_object[["mito.genes"]] <- CCA_object_subset$mito.genes
multimodal_object[["ribo.genes"]] <- CCA_object_subset$ribo.genes
multimodal_object[["dataset"]] <- LSI_object_subset$dataset

# and reductions
multimodal_object@reductions$integrated_pca <- CCA_object_subset@reductions$pca
multimodal_object@reductions$integrated_umap <- CCA_object_subset@reductions$umap
multimodal_object@reductions$integrated_lsi <- LSI_object_subset@reductions$integrated_lsi
multimodal_object@reductions$lsi_umap <- LSI_object_subset@reductions$umap

# add scaled data
all.genes <- rownames(multimodal_object)
multimodal_object <- ScaleData(multimodal_object, features = all.genes, assay = "RNA")

DefaultAssay(multimodal_object) <- "ATAC"
all.peaks <- rownames(multimodal_object)
#multimodal_object <- ScaleData(multimodal_object, features = all.peaks, assay = "ATAC")
#multimodal_object <- NormalizeData(multimodal_object, assay = "RNA")

# add the gene activity matrix to the Seurat object as a new assay and normalize it
#gene.activities <- GeneActivity(multimodal_object, assay = "ATAC")
#multimodal_object[['gene_activity']] <- CreateAssayObject(counts = gene.activities)
#multimodal_object <- NormalizeData(
#  object = multimodal_object,
#  assay = 'gene_activity',
#  normalization.method = 'LogNormalize',
#  scale.factor = median(multimodal_object$nCount_gene_activity)
#)

#DefaultAssay(multimodal_object) <- "gene_activity"
#all.ga <- rownames(multimodal_object)
#multimodal_object <- ScaleData(multimodal_object, features = all.ga, assay = "gene_activity")

# run WNN using reductions from RNA and ATAC
multimodal_object <- FindMultiModalNeighbors(multimodal_object, 
                                             reduction.list = list("integrated_pca",
                                                                   "integrated_lsi"), 
                                             dims.list = list(1:30, 2:50))

# Run UMAP and find clusters in multimodal data
multimodal_object <- RunUMAP(multimodal_object, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
multimodal_object <- FindClusters(multimodal_object, graph.name = "wsnn", algorithm = 3, verbose = FALSE, resolution = 0.5)

DimPlot(multimodal_object, reduction = "wnn.umap", pt.size = 0.1)
```

##################################################
############## ANNOTATION OF OBJECT ##############
##################################################
```{r}
# add  annotations
disease <- ifelse(
  str_detect(
    multimodal_object[["dataset"]][["dataset"]], "21_016"
    ), "MELAS", "Control")


multimodal_object[["disease"]] <- disease


# plot outcomes of WNN analysis
  # load rpe/chor colors!
p1 <- DimPlot(multimodal_object, reduction = "integrated_umap", group.by = "wsnn_res.0.5", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("RNA")
p2 <- DimPlot(multimodal_object, reduction = "lsi_umap", group.by = "wsnn_res.0.5", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC")
p3 <- DimPlot(multimodal_object, reduction = "wnn.umap", group.by = "wsnn_res.0.5", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")



pdf(file = "melas_wnn_triple_retina_01112023.pdf",   # The directory you want to save the file in
    width = 15, # The width of the plot in inches
    height = 5) # The height of the plot in inches

p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))

dev.off()

DefaultAssay(multimodal_object) <- "RNA"
FeaturePlot(multimodal_object, reduction = "wnn.umap",
            features = "RGCC", order = T,
            max.cutoff = "q95", min.cutoff = "q05")

DimPlot(multimodal_object, split.by = "disease", reduction = "wnn.umap")

FeaturePlot(subset(multimodal_object, cov3243 > 10),  reduction = "wnn.umap",
            features = "het3243A_G", order = F, split.by = "dataset",
            cols = c("blue", "white", "red"), pt.size = 0.1, ncol = 5)


DefaultAssay(multimodal_object) <- "RNA"
FeaturePlot(multimodal_object, reduction = "wnn.umap",
            features = "het3243A_G", cols = c("blue", "red"), order = T)


VlnPlot(multimodal_object, features = "cov3243", group.by = "dataset", log = T, pt.size = 0.1)
```




```{r}

donor_wnn_dim <- DimPlot(multimodal_object,shuffle = T, pt.size = 0.1, group.by = "dataset", reduction = "wnn.umap") + ggtitle("Retina Multimodal Seq")
donor_split_wnn_dim <- DimPlot(multimodal_object,shuffle = T, pt.size = 0.1, split.by  = "line", reduction = "wnn.umap") + ggtitle("D160 RO Multimodal Seq")


pdf(file = "donor_wnn_dimplot_retina_01112023.pdf",   # The directory you want to save the file in
    width = 20, # The width of the plot in inches
    height = 5) # The height of the plot in inches
donor_wnn_dim 
dev.off()

pdf(file = "donor_split_wnn_dimplot_12132022.pdf",   # The directory you want to save the file in
    width = 10, # The width of the plot in inches
    height = 8) # The height of the plot in inches
donor_split_wnn_dim 
dev.off()

DefaultAssay(multimodal_object) <- "RNA"

pdf(file = "rom1_pde6h.pdf",   # The directory you want to save the file in
    width = 15, # The width of the plot in inches
    height = 8) # The height of the plot in inches
FeaturePlot(multimodal_object, pt.size = 0.1, split.by  = "line", reduction = "wnn.umap", 
            features = c("ROM1", "PDE6H"), cols = c("red", "blue"), blend.threshold = 0.25, order = F, blend = T)
dev.off()


pdf(file = "rho_nrl.pdf",   # The directory you want to save the file in
    width = 15, # The width of the plot in inches
    height = 20) # The height of the plot in inches
FeaturePlot(multimodal_object, pt.size = 0.1, reduction = "wnn.umap", 
            features = c("NRL", "ARR3", "RHO", "ROM1"), max.cutoff = "q95", min.cutoff = "q05", order = F)
dev.off()

Split_FeatureScatter(seurat_object = multimodal_object, feature1 = "ROM1", feature2 = "PDE6H", split.by = "line")

```

```{r}
save(multimodal_object, file = "multimodal_object_retina_01142023.RData")
```


NOW ANNOTATE CELL TYPE USING NR2E3 DATA (?)

###########################
#### Reference Mapping ####
###########################

```{r}
# load NR2E3 RO data
load("~/LSS/IVR/nkmullin/experiments/sc-rna-seq/melas_jci_insight_revision/celltype_annotation_of_rna/retina_bullseye_controls_annotated_11012022.RData")

DefaultAssay(IVR_object.combined) <- "RNA"
DefaultAssay(multimodal_object) <- "RNA"

common.features <- intersect(rownames(multimodal_object), rownames(IVR_object.combined))
length(x = common.features)

anchors <- FindTransferAnchors(reference = IVR_object.combined, query = multimodal_object,
    dims = 1:30, reference.reduction = "pca", features = common.features, reduction = "pcaproject")

predictions <- TransferData(anchorset = anchors, refdata = IVR_object.combined$celltype,
    dims = 1:30)


multimodal_object <- AddMetaData(multimodal_object, metadata = predictions)
Idents(multimodal_object) <- 'predicted.id'

pdf("retina_pred_scor_01122023.pdf", width=5, height=5)
FeaturePlot(multimodal_object, features =  "prediction.score.max", reduction = "wnn.umap")
dev.off()

# set really low prediction score cutoff ie zero!
multimodal_object$draft_celltype <- ifelse(multimodal_object$prediction.score.max > 0.5, multimodal_object$predicted.id, NA)



Idents(multimodal_object) <- 'draft_celltype'

pdf("dimplot_retina_draft_annotation_01122023.pdf", width=5, height=5)
DimPlot(multimodal_object, reduction = "wnn.umap", shuffle = T, label = T, group.by = "draft_celltype", ncol = 1, cols = c(brewer.pal(n = 10, name = "Set1"), 
brewer.pal(n = 10, name = "Dark2"))) + NoLegend()
dev.off()


table(multimodal_object$dataset, multimodal_object$draft_celltype)

VlnPlot(multimodal_object, group.by = "dataset", split.by = "draft_celltype", features = "het3243A_G", pt.size = 0)

multimodal_object

sub <- subset(multimodal_object, cov3243 > 10)

VlnPlot(sub, group.by = "dataset", split.by = "draft_celltype", features = "het3243A_G", pt.size = 0)

DimPlot(sub, split.by = "dataset")

save(multimodal_object, file = "multimodal_object_retina_01122023_draft_annotation.RData")


DefaultAssay(multimodal_object) <- "ATAC"


CoveragePlot(obj, features = "RHO", region = "RHO")

obj = subset(multimodal_object, draft_celltype != "NA")

cov_plot <- CoveragePlot(
  object = obj,
  region = "chr2-87011729-87035519",
  annotation = FALSE,
  peaks = FALSE
)
cov_plot


CoveragePlot(obj, features = "NRL", region = "NRL")


peaks <- CallPeaks(obj, macs2.path = "~/.local/bin/macs3")

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(obj),
  features = peaks,
  cells = colnames(obj)
)

# create a new assay using the MACS2 peak set and add it to the Seurat object
obj[["macs_peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = obj@assays$ATAC@fragments,
  annotation = annotations
)

DefaultAssay(obj) <- "macs_peaks"



```
